### ISSUE TITLE
Indexing into arrays hoists empty array constants out of Jaxprs

### ISSUE THREAD

**Issue by ahoenselaar** [google/jax#6404]
Repro:
```
def test_fun(x):
  return x[0] + x[1] + x[2] + x[3]
```

Jaxpr:
```
{ lambda a b c d ; e.
  let f = lt 0 0
      g = add 0 4
      h = select f g 0
      i = convert_element_type[ new_dtype=int32
                                weak_type=False ] h
      j = broadcast_in_dim[ broadcast_dimensions=(  )
                            shape=(1,) ] i
      k = concatenate[ dimension=0 ] a j
      l = gather[ dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))
                  slice_sizes=(1,) ] e k
  [...]
}
```
where `a`, `b`, `c`, and `d` are empty int32 arrays - one per indexing operation: `[array([], dtype=int32), array([], dtype=int32), array([], dtype=int32), array([], dtype=int32)]`

In severe cases, thousands of these constants inflate the size of the unoptimized HLO and result in slow compilation. In code with higher-order functions, there is evidence that these constants and the associated `lt`,`add`, and `select` operations do not get optimized away.

Workarounds: `jnp.take` does not create empty constants and `lax.slice` results in even more compact code in situations where its semantics are applicable.

**Comment by hawkinsp** [comment]
With #6416 

```
In [1]: import jax

In [2]: jax.make_jaxpr(lambda x: x[2])(jax.numpy.arange(5))
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
Out[2]:
{ lambda  ; a.
  let b = lt 2 0
      c = add 2 5
      d = select b c 2
      e = broadcast_in_dim[ broadcast_dimensions=(  )
                            shape=(1,) ] d
      f = gather[ dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))
                  slice_sizes=(1,) ] a e
  in (f,) }

```

**Comment by hawkinsp** [comment]
I am not completely confident this will fix your original problem, but I'm hopeful that it will. Please let me know!

**Comment by ahoenselaar** [comment]
This looks a lot better, thanks! We are now lowering directly to `lax.index_in_dim` to take advantage of the Slice XLA op.

**Comment by hawkinsp** [comment]
That's an optimization we can make also, but I am yet to see evidence that it helps. Do you have an example where it matters?
